src/physfs_platform_libretro.c

```diff
@@ -0,0 +1,421 @@
+/*
+ * libretro support routines for PhysicsFS.
+ *
+ * Please see the file LICENSE.txt in the source's root directory.
+ *
+ * This file written by Vladimir Serbinenko (@phcoder), and Rob Loach (@RobLoach).
+ */
+
+#define __PHYSICSFS_INTERNAL__
+#include "physfs_platforms.h"
+
+#ifdef PHYSFS_PLATFORM_LIBRETRO
+
+#include <string.h>
+#include <sys/types.h>
+#include <sys/stat.h>
+#include <unistd.h>
+#include <errno.h>
+
+/* libretro dependencies */
+#include <libretro.h>
+#include <rthreads/rthreads.h>
+
+#include "physfs_internal.h"
+
+static retro_environment_t physfs_platform_libretro_environ_cb = NULL;
+static struct retro_vfs_interface *physfs_platform_libretro_vfs = NULL;
+
+
+int __PHYSFS_platformInit(const char *argv0)
+{
+    /* as a cheat, we expect argv0 to be a retro_environment_t callback. */
+    retro_environment_t environ_cb = (retro_environment_t) argv0;
+    struct retro_vfs_interface_info vfs_interface_info;
+    BAIL_IF(environ_cb == NULL || argv0[0] == '\0', PHYSFS_ERR_INVALID_ARGUMENT, 0);
+
+    /* get the virtual file system interface, at least version 3 */
+    vfs_interface_info.required_interface_version = 3;
+    vfs_interface_info.iface = NULL;
+    BAIL_IF(!environ_cb(RETRO_ENVIRONMENT_GET_VFS_INTERFACE, &vfs_interface_info), PHYSFS_ERR_UNSUPPORTED, 0);
+
+    physfs_platform_libretro_vfs = vfs_interface_info.iface;
+    physfs_platform_libretro_environ_cb = environ_cb;
+
+    return 1;
+} /* __PHYSFS_platformInit */
+
+
+void __PHYSFS_platformDeinit(void)
+{
+    physfs_platform_libretro_environ_cb = NULL;
+    physfs_platform_libretro_vfs = NULL;
+} /* __PHYSFS_platformDeinit */
+
+
+void __PHYSFS_platformDetectAvailableCDs(PHYSFS_StringCallback cb, void *data)
+{
+} /* __PHYSFS_platformDetectAvailableCDs */
+
+
+char *physfs_platform_libretro_get_directory(int libretro_dir, int add_slash)
+{
+    const char *dir = NULL;
+    char *retval;
+    size_t dir_length;
+
+    BAIL_IF(physfs_platform_libretro_environ_cb == NULL, PHYSFS_ERR_NOT_INITIALIZED, NULL);
+    BAIL_IF(!physfs_platform_libretro_environ_cb(libretro_dir, &dir), PHYSFS_ERR_UNSUPPORTED, NULL);
+
+    dir_length = strlen(dir);
+    retval = allocator.Malloc(dir_length + 2);
+    if (!retval) {
+        BAIL(PHYSFS_ERR_OUT_OF_MEMORY, NULL);
+    }
+
+    strcpy(retval, dir);
+
+    if (add_slash && (dir_length == 0 || retval[dir_length - 1] != __PHYSFS_platformDirSeparator)) {
+        retval[dir_length] = __PHYSFS_platformDirSeparator;
+        retval[dir_length + 1] = '\0';
+    }
+
+    return retval;
+}
+
+
+char *__PHYSFS_platformCalcBaseDir(const char *argv0)
+{
+    /* TODO: Replace getcwd() with a libretro-common function call? */
```

Comment:
For the base directory, it's currently using `getcwd()`, but this may not be the best solution. Maybe use `RETRO_ENVIRONMENT_GET_LIBRETRO_PATH` instead?

Reply:
This is meant to be where to find game data files...which for a libretro core is either going to be the engine's data or the game/rom-specific data...I'd lean towards the latter, if they aren't the same thing.

Reply:
Good brainstorming, thanks. Made an update to use....
1. Try to get the [game directory via `GET_GAME_INFO_EXT`](https://github.com/libretro/RetroArch/blob/master/libretro-common/include/libretro.h#L2266-L2268), which is a [directory in which the content file exists](https://github.com/libretro/RetroArch/blob/master/libretro-common/include/libretro.h#L6141-L6147)
2. Otherwise, use [GET_SYSTEM_DIRECTORY](https://github.com/libretro/RetroArch/blob/master/libretro-common/include/libretro.h#L828-L844), as that may make sense for data files
3. If both fail, use `/`